landoapi/hgexports.py

```diff
@@ -605,3 +616,144 @@ def run_diff_checks(self) -> list[str]:
+
+
+@dataclass
+class PushCheck:
+    """Provides an interface to implement push-wide checks."""
+
+    repo: Repo
+
+    def relevant(self) -> bool:
+        """Return `True` if the check is relevant for the given repo."""
+        raise NotImplementedError()
+
+    def next_diff(self, patch_helper: PatchHelper):
+        """Pass the next diff to be checked into the check and run."""
+        raise NotImplementedError()
+
+    def result(self) -> Optional[str]:
+        """Return the result of the check."""
+        raise NotImplementedError()
+
+
+@dataclass
+class BugReferencesCheck(PushCheck):
+    """Prevent commit messages referencing non-public bugs from try."""
+
+    bug_ids: set[int] = field(default_factory=set)
+    skip_check: bool = False
+
+    def relevant(self) -> bool:
+        if self.repo.tree == "try":
+            return True
+
+        return False
+
+    def next_diff(self, patch_helper: PatchHelper):
+        commit_message = patch_helper.get_commit_description()
+
+        # Skip the check if the `skip_check` flag is set.
+        if self.skip_check or "SKIP_BMO_CHECK" in commit_message:
+            self.skip_check = True
+            return
+
+        self.bug_ids |= set(parse_bugs(commit_message))
+
+    def result(self) -> Optional[str]:
+        if self.skip_check or not self.bug_ids:
+            return
+
+        try:
+            found_bugs = search_bugs(self.bug_ids)
+        except requests.exceptions.RequestException as exc:
+            return (
+                "Could not contact BMO to check for security bugs referenced in commit message. "
+                f"Use `SKIP_BMO_CHECK` in your commit message to push anyways. Error: {str(exc)}."
+            )
+
+        invalid_bugs = self.bug_ids - found_bugs
+        if not invalid_bugs:
+            return
+
+        # Check a single bug only.
+        bug_id = invalid_bugs.pop()
+        try:
+            status_code = get_status_code_for_bug(bug_id)
+        except requests.exceptions.RequestException as exc:
+            return (
+                "Could not contact BMO to check for security bugs referenced in commit message. "
+                f"Use `SKIP_BMO_CHECK` in your commit message to push anyways. Error: {str(exc)}."
+            )
+
+        if status_code == 401:
+            return (
+                f"Your commit message references bug {bug_id}, which is currently private. To avoid "
+                "disclosing the nature of this bug publicly, please remove the affected bug ID "
+                "from the commit message."
+            )
+
+        if status_code == 404:
+            return (
+                f"Your commit message references bug {bug_id}, which does not exist. Please check your "
+                "commit message and try again."
+            )
+
+        return (
+            f"While checking if bug {bug_id} in your commit message is a security bug, an error occurred "
+            "and the bug could not be verified."
+        )
```

Comment:
Would be cleaner to put all the error strings into a dictionary and pass an "error code" to fetch the right string here, instead of scattering the strings across this method. Error codes could look like `ERROR_BUGZILLA_UNAVAILABLE`, or `ERROR_BUG_DOES_NOT_EXIST`.

Reply:
What would make that approach cleaner? I'd rather not move the error messages away from the code that is triggering them. Also it seems like an unnecessary abstraction because the messages are only used in 1-2 locations within a single function.